### PR TITLE
R script for mzTab to csv conversion

### PR DESCRIPTION
**Pull request by lars20070**
The R script generates from the mzTab input a list of unique modified peptide sequences and their attributes. Only the peptide sequence with the highest posterior error probability is kept. The output is a simple csv table with a single header.

### CODE REVIEW

**Comment by timosachsenberg** [comment]
the script is very specific for your special use case and not as generic as the name might imply. I am not sure if we really want to include it in its current form.

**Comment by hroest** [comment]
I am personally not really against having a place for such R scripts, especially if they are used in a labs standard pipeline and are not one-off scripts. I think there should be a place to put those scripts but I am also not sure where the best place would be for them. I think the most important point here should be that there is no official support for these scripts so even if we agree to collect them in this repo they are basically "use at own risk". @timosachsenberg the other scripts in here also seem to be very specific for a usecase and not very generic https://github.com/OpenMS/OpenMS/tree/develop/share/OpenMS/SCRIPTS

Another option would be to have a new git project inside the OpenMS page, similar to the contrib project, where such scripts can be collected. Maybe this would make the separation from the "official" OpenMS more clear while still indicating that these scripts are intended for the data files written by OpenMS

**Comment by timosachsenberg** [comment]
@mwalzer what is the state of the https://github.com/OpenMS/OpenMS/tree/develop/share/OpenMS/SCRIPTS scripts? Are they still needed?

**Comment by lars20070** [comment]
From the command line, teh script runs fine.
`Rscript --vanilla OpenMS/share/OpenMS/SCRIPTS/mzTab2csv_peptides.R WL383.tsv WL383.csv`

Yet, within TOPPAS the GenericWrapper still crashes.
`13:46:18 NOTICE: GenericWrapper (Rscript_mzTab2csv_peptides) of node #2 started. Processing ...
Warning: Unable to fetch subsection parameters! Addressing subsection parameters will not work for this tool.
Calling Rscript mzTab2csv_peptides ...
GenericWrapper took 0.00 s (wall), 0.00 s (CPU), 0.00 s (system), 0.00 s (user).
External tool returned with non-zero exit code (0), exit status (0) or timed out. Aborting ...`

Seems the `ETool` subsection params, i.e. the script location (`file_script`), are not found.

**Comment by cbielow** [comment]
call GW with '-debug 99' and look at the command line.
It is executed in the working directory where TOPPAS was launched from.
So if the path to the script is relative, this might be the reason.

**Comment by lars20070** [comment]
Warning: Unable to fetch subsection parameters! Addressing subsection parameters will not work for this tool.
Debug level: 99
GenericWrapper -debug 99 -ini GW.ini -in WL383.tsv -out WL383.csv
Unknown option(s) '[-out, -in]' given. Aborting!
Parameter '-helphelp' not found.
Value of string option '-helphelp': 0

GenericWrapper -- Allows the generic wrapping of external tools.
Version: 2.0.0 Jun  8 2015, 14:18:08, Revision: 982ea6d

**Comment by lars20070** [comment]
`Unknown option(s) '[-out, -in]' given.`
Seems input and output are not found.

**Comment by cbielow** [comment]
No,
your call is wrong.
Should be 
GenericWrapper -debug 99 -ini GW.ini -ETool:in WL383.tsv -ETool:out WL383.csv

To be sure, use TOPPAS :)

**Comment by lars20070** [comment]
Enqueue: "/Users/lars/Code/git/OpenMS/bin/TOPPAS.app/Contents/MacOS/GenericWrapper" "-type" "Rscript_mzTab2csv_peptides" "-ini" "/var/folders/1r/lb4bxs4d2gx93zsb18_b65100000gn/T/2015-06-08_142535_immz1179.molmed.uni-freiburg.de_15059_1/TOPPAS_tmp/mzTab2csv/002_GenericWrapper_Rscript_mzTab2csv_peptides/GenericWrapper_Rscript_mzTab2csv_peptides0.ini"

Enqueue: "/Users/lars/Code/git/OpenMS/bin/TOPPAS.app/Contents/MacOS/GenericWrapper" "-type" "Rscript_mzTab2csv_peptides" "-ini" "/var/folders/1r/lb4bxs4d2gx93zsb18_b65100000gn/T/2015-06-08_142535_immz1179.molmed.uni-freiburg.de_15059_1/TOPPAS_tmp/mzTab2csv/002_GenericWrapper_Rscript_mzTab2csv_peptides/GenericWrapper_Rscript_mzTab2csv_peptides0.ini"

==============================================================================
14:37:42 NOTICE: GenericWrapper (Rscript_mzTab2csv_peptides) of node #2 started. Processing ...
Warning: Unable to fetch subsection parameters! Addressing subsection parameters will not work for this tool.
Debug level (after ini file): 99
Value of string option 'no_progress': 1
Value of string option 'type': Rscript_mzTab2csv_peptides
Calling Rscript mzTab2csv_peptides ...
CommandLine from ttd (unprocessed): --vanilla %1 %2 %3
call command: Rscript --vanilla /Users/lars/Code/git/OpenMS/share/OpenMS/SCRIPTS/mzTab2csv_peptides.R /Users/lars/Desktop/WL383.tsv /var/folders/1r/lb4bxs4d2gx93zsb18_b65100000gn/T/2015-06-08_142535_immz1179.molmed.uni-freiburg.de_15059_1/TOPPAS_tmp/mzTab2csv/002_GenericWrapper_Rscript_mzTab2csv_peptides/EToolout/WL383.tsv_tmp1
GenericWrapper took 0.00 s (wall), 0.00 s (CPU), 0.00 s (system), 0.00 s (user).
External tool returned with non-zero exit code (0), exit status (0) or timed out. Aborting ...
External tool output:

Something went wrong. Is the 'Rscript' executable in your PATH?

**Comment by lars20070** [comment]
The `call command` looks fine to me.

**Comment by cbielow** [comment]
so what happens if you copy the command and execute it manually?
(keep TOPPAS etc open, so that the intermediate input files  -if any - 
are preserved)

Am 6/8/2015 um 2:39 PM schrieb Lars Nilse:

**Comment by lars20070** [comment]
Executing the toppas workflow, the tmp folder is generated, but it contains no result.

Running the copy/pasted command line, the WL383.tsv_tmp1 result is generated. No errors. R script warnings are switched off.

**Comment by lars20070** [comment]
Here a example file, mzTab input and csv output.
https://www.dropbox.com/sh/egb9maq6jox608t/AAAwyeyt10zKwEJ5bHTQDcxDa

File: share/OpenMS/TOOLS/EXTERNAL/Rscript_mzTab2csv_peptides.ttd [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+			<e_category>Postprocessing</e_category>
+			<cloptions>--vanilla %1 %2 %3</cloptions>
+			<path>Rscript</path>
+			<mappings>
+				<mapping id="1" cl="%%file_script" />
+				<mapping id="2" cl="%%in" />
+				<mapping id="3" cl="%%out" />
+			</mappings>
+			<ini_param>
+				<ITEM name="file_script" value="" type="string" description="R script" tags="input file" />
```
Comment (cbielow) [inline share/OpenMS/TOOLS/EXTERNAL/Rscript_mzTab2csv_peptides.ttd]: this should be
```
<ITEM name="file_script" value="" type="input-file" description="R script" required="true" advanced="false" />
```
i.e. new Param format and required=true

**Comment by cbielow** [comment]
The above GW node runs perfectly fine on my system (Win7 x64, R 3.1.2 x64).

**Comment by lars20070** [comment]
GW with my specific R script fails on OS X and Windows Server 2012.

**Comment by mwalzer** [comment]
@timosachsenberg scripts are not some old deprecated stuff but used for building KNIME QC subworkflows.

**Comment by lars20070** [comment]
FYI. As discussed with @cbielow by email, the problem above seems to be an OS X QProcess problem and not related to the ttd file.

File: share/OpenMS/SCRIPTS/mzTab2csv_peptides.R [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	
+	indices <- c(indices, k)
+}
+
+# simplify accession
+list <- strsplit(peptide.data$accession,"[|]")
+peptide.data$accession <- unlist(lapply(list, '[[', 2))
+
+# add fold change
+peptide.data$fold_change <- log(as.numeric(peptide.data$"peptide_abundance_study_variable[2]")/as.numeric(peptide.data$"peptide_abundance_study_variable[1]"), base =2)
```
Comment (timosachsenberg) [inline share/OpenMS/SCRIPTS/mzTab2csv_peptides.R]: ony works if two study variables

File: share/OpenMS/SCRIPTS/mzTab2csv_peptides.R [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+unique.stripped.sequences <- unique(peptide.data$sequence)
+unique.modified.sequences <- unique(peptide.data$opt_global_modified_sequence)
+
+# find best (lowest PEP score) modified sequences
+n <- length(unique.modified.sequences)
+indices <- c()
+for (i in 1:n) {
+	j <- which(peptide.data$"opt_global_modified_sequence" == unique.modified.sequences[i])
+	PEPs <- peptide.data[j,]$"opt_psm_Posterior Error Probability_score"
```
Comment (timosachsenberg) [inline share/OpenMS/SCRIPTS/mzTab2csv_peptides.R]: probably only works if column present

**Comment by lars20070** [comment]
The pull request is updated, and now contains three separate `mzTab2tsv` scripts for the `PRT`, `PEP` and `PSM` sections of `mzTab`.

**Comment by lars20070** [comment]
Comments? Merge?

**Comment by timosachsenberg** [comment]
Hi Lars, can you post a gist for the output of the different scripts?

**Comment by lars20070** [comment]
Here mzTab and corresponding tsv with PRT section [1,2]. And here with a PEP section [3,4].

[1] https://gist.github.com/lars20070/05313e57202567011887
[2] https://gist.github.com/lars20070/f327f17d971192e50a84
[3] https://gist.github.com/lars20070/7cc65fcb5a16c4aa4972
[4] https://gist.github.com/lars20070/499dcf94a98720b2f391

**Comment by timosachsenberg** [comment]
If a column is completly "null" one could remove it I guess (its only mandatory in mzTab).

**Comment by lars20070** [comment]
Yep, easily done. - For now, it makes debugging the OpenMS mzTab export easier. All columns in mzTab appear in the tsv.

**Comment by timosachsenberg** [comment]
what is the result if there are no protein groups? does it still work?

**Comment by timosachsenberg** [comment]
what if the section is missing? seems to produce invalid results looking at the code?

**Comment by lars20070** [comment]
Good point. Thanks. - If no groups are reported in PRT, I will fall back on a trivial output as in PEP and PSM.

**Comment by lars20070** [comment]
@timosachsenberg The PRT script now checks for protein group information.

**Comment by lars20070** [comment]
Just a time out. Can we merge these scripts?

**Comment by timosachsenberg** [comment]
but please keep an eye on recent developments as these are not automatically tested

**Pull request merged**